src/backend/distributed/utils/multi_partitioning_utils.c

```diff
@@ -130,6 +139,124 @@ worker_fix_pre_citus10_partitioned_table_constraint_names(PG_FUNCTION_ARGS)
+ *
+ * fix_partition_shard_index_names renames indexes of shards of partition tables to include
+ * the shardId at the end of the name, regardless of whether index name was long or short
+ * As a result there will be no index name ending in _idx, rather all will end in _{shardid}
+ */
+Datum
+fix_partition_shard_index_names(PG_FUNCTION_ARGS)
+{
+	Oid relationId = PG_GETARG_OID(0);
+	EnsureCoordinator();
+
+	Relation relation = try_relation_open(relationId, AccessShareLock);
+
+	if (relation == NULL)
+	{
+		ereport(ERROR, (errmsg("could not fix partition shard index names: "
+							   "relation does not exist")));
+	}
+
+	if (relation->rd_rel->relkind != RELKIND_PARTITIONED_TABLE)
+	{
+		relation_close(relation, AccessShareLock);
+		ereport(ERROR, (errmsg("could not fix partition shard index names: "
+							   "relation is not partitioned")));
+	}
+
+	if (!IsCitusTable(relationId))
+	{
+		relation_close(relation, AccessShareLock);
+		ereport(ERROR, (errmsg("fix_partition_shard_index_names can "
+							   "only be called for distributed partitioned tables")));
+	}
+
+	List *taskList = CreateFixPartitionShardIndexNamesTaskList(relationId);
+
+	/* do not do anything if there are no index names to fix */
+	if (taskList != NIL)
+	{
+		bool localExecutionSupported = true;
+		ExecuteUtilityTaskList(taskList, localExecutionSupported);
+	}
+
+	relation_close(relation, AccessShareLock);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * worker_fix_partition_shard_index_names fixes the index name of the index on given
+ * partition shard that has parent the given parent index.
+ * The parent index should be the index of a shard of a distributed partitioned table.
+ */
+Datum
+worker_fix_partition_shard_index_names(PG_FUNCTION_ARGS)
+{
+	Oid shardIndexId = PG_GETARG_OID(0);
+
+	text *partitionShardName = PG_GETARG_TEXT_P(1);
+
+	/* resolve partitionShardId from passed in schema and partition shard name */
+	List *partitionShardNameList = textToQualifiedNameList(partitionShardName);
+	RangeVar *partitionShard = makeRangeVarFromNameList(partitionShardNameList);
+
+	/* lock the relation with the lock mode */
+	Oid partitionShardId = RangeVarGetRelid(partitionShard, NoLock, RVR_MISSING_OK);
+
+	if (OidIsValid(partitionShardId))
+	{
+		text *newPartitionShardIndexNameText = PG_GETARG_TEXT_P(2);
+		char *newPartitionShardIndexName = text_to_cstring(
+			newPartitionShardIndexNameText);
+
+		if (!has_subclass(shardIndexId))
+		{
+			ereport(ERROR, (errmsg("could not fix child index names: "
+								   "index is not partitioned")));
+		}
+
+		List *partitionShardIndexIds = find_inheritance_children(shardIndexId,
+																 ShareRowExclusiveLock);
+		Oid partitionShardIndexId = InvalidOid;
+		foreach_oid(partitionShardIndexId, partitionShardIndexIds)
+		{
+			if (IndexGetRelation(partitionShardIndexId, false) == partitionShardId)
+			{
+				char *partitionShardIndexName = get_rel_name(partitionShardIndexId);
+				if (ExtractShardIdFromTableName(partitionShardIndexName,
```

Comment:
Maybe improve the comment below a little more. Why do we only rename for non-valid shard ids?